devito/finite_differences/finite_difference.py

```diff
@@ -81,6 +82,15 @@ def second_derivative(expr, **kwargs):
        results in ``(-2.0*f(x, y)*g(x, y) + 1.0*f(-h + x, y)*g(-h + x, y) +
        1.0*f(h + x, y)*g(h + x, y)) / h**2``.
     """
+    # Check if we need symbolic fd weights
+    functions = expr.as_ordered_factors()
```

Comment:
I suppose here you want the actual {Time}Function objects, while ``as_ordered_factors`` gives you...the factors, which can actually be generic expressions. And I assume you don't care about repetitions, ie if you had ``f(x+h_x, y), f(x, y), f(x, y+h_y)`` retrieving ``f(x, y)`` (xor ``f(x+h_x, y)`` xor ...)) would suffice.

If that's the case, you can:
```
from devito.symbolics import retrieve_functions
functions = retrieve_functions(expr)  # get all {Time,Sparse,...}Function objects

from devito.tools import filter_ordered
functions = filter_ordered(functions, key=lambda i: i.name)
```

Even better: ``retrieve_functions`` takes in input the optional parameter mode. Currently, ``mode`` can be either 'unique' or 'all'. Don't get confused by the word 'unique' here: it simply means that if in your expression ``f(x, y)`` appears twice, then you only fetch it once. But ``f(x, y)`` is still different than ``f(x+h-x, y)``. IOW, here we need uniqueness based on the Function name only, not on name+indices. So, the better way to go is add a new mode.

Reply:
Yes, will think a little more, but I believe we just wan't to pull unique functions. So will look into adding a new mode.